src/main/java/uk/gov/ons/census/casesvc/service/QuestionnaireLinkedService.java

```diff
@@ -50,10 +52,25 @@ public void processQuestionnaireLinked(
 
     Case caze = caseService.getCaseByCaseId(UUID.fromString(uac.getCaseId()));
 
-    if (isIndividualQuestionnaireType(questionnaireId) && caze.getCaseType().equals("HH")) {
-      caze = caseService.prepareIndividualResponseCaseFromParentCase(caze, UUID.randomUUID());
+    if (checkRequiredFieldsForIndividualHI(
+        questionnaireId, caze.getCaseType(), uac.getIndividualCaseId())) {
+      caze =
+          caseService.prepareIndividualResponseCaseFromParentCase(
+              caze, UUID.fromString(uac.getIndividualCaseId()));
       caze = caseService.saveNewCaseAndStampCaseRef(caze);
       caseService.emitCaseCreatedEvent(caze);
+    } else if (checkIfIndQIDAndIndCaseIdPresentButCaseTypeNotHH(
```

Comment:
instead of 2 if elses, could there be 2 function calls both of which just throw the exception? maybe renamed to reflect that? Not sure if that would be great practice or not, but would stop the 'crowding' in the main function.

```suggestion
       if (checkRequiredFieldsForIndividualHI(questionnaireId, caze, uac.getIndividualCaseId())) {
      caze =
          caseService.prepareIndividualResponseCaseFromParentCase(
              caze, UUID.fromString(uac.getIndividualCaseId()));
      caze = caseService.saveNewCaseAndStampCaseRef(caze);
      caseService.emitCaseCreatedEvent(caze);
    } else {
      checkIfIndQIDAndIndCaseIdPresentButCaseTypeNotHH(
          questionnaireId, caze, uac.getIndividualCaseId());
      checkIfIndQidAndCaseTypeHHButIndCaseIdNotPresent(
          questionnaireId, caze, uac.getIndividualCaseId());
    }

    uacQidLink.setCaze(caze);

    if (!uacQidLink.isActive()) {
      if (uacQidLink.isBlankQuestionnaire()) {
        blankQuestionnaireService.handleBlankQuestionnaire(
            caze, uacQidLink, questionnaireLinkedEvent.getEvent().getType());
      } else {
        caseReceiptService.receiptCase(uacQidLink, questionnaireLinkedEvent.getEvent().getType());
      }
    }

    uacService.saveAndEmitUacUpdatedEvent(uacQidLink);

    eventLogger.logUacQidEvent(
        uacQidLink,
        questionnaireLinkedEvent.getEvent().getDateTime(),
        QUESTIONNAIRE_LINKED,
        EventType.QUESTIONNAIRE_LINKED,
        questionnaireLinkedEvent.getEvent(),
        convertObjectToJson(uac),
        messageTimestamp);
  }

  private void logUnlinkedCaseEventAgainstOriginalCaseIfQidIsNowLinkedToNewCase(
      OffsetDateTime messageTimestamp,
      ResponseManagementEvent questionnaireLinkedEvent,
      UacDTO uac,
      UacQidLink uacQidLink) {
    if (uacQidLink.getCaze() != null
        && !uacQidLink.getCaze().getCaseId().equals(UUID.fromString(uac.getCaseId()))) {
      String eventDescription =
          String.format("Questionnaire unlinked from case with QID %s", uac.getQuestionnaireId());
      eventLogger.logCaseEvent(
          uacQidLink.getCaze(),
          questionnaireLinkedEvent.getEvent().getDateTime(),
          eventDescription,
          EventType.QUESTIONNAIRE_UNLINKED,
          questionnaireLinkedEvent.getEvent(),
          convertObjectToJson(uac),
          messageTimestamp);
    }
  }

  private boolean checkRequiredFieldsForIndividualHI(
      String questionnaireId, Case caze, String individualCaseId) {
    return (isIndividualQuestionnaireType(questionnaireId)
        && caze.getCaseType().equals("HH")
        && individualCaseId != null);
  }

  private void checkIfIndQIDAndIndCaseIdPresentButCaseTypeNotHH(
      String questionnaireId, Case caze, String individualCaseId) {
    if (isIndividualQuestionnaireType(questionnaireId)
        && !caze.getCaseType().equals("HH")
        && individualCaseId != null) {
      throw new IllegalArgumentException(
          String.format(
              "CaseType on '%s' not HH where QID is individual on PQ link request where individualCaseId provided",
              caze.getCaseId()));
    }
  }

  private void checkIfIndQidAndCaseTypeHHButIndCaseIdNotPresent(
      String questionnaireId, Case caze, String individualCaseId) {
    if (isIndividualQuestionnaireType(questionnaireId)
        && caze.getCaseType().equals("HH")
        && isEmpty(individualCaseId)) {
      throw new IllegalArgumentException(
          String.format(
              "No individualCaseId present on PQ link request where QID is individual and caseType for '%s' is HH",
              caze.getCaseId()));
    }
  }
```

Reply:
Made some changes to more of this pattern 👍